In [1]:
import pandas as pd
import numpy as np

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    confusion_matrix,
    classification_report
)

In [2]:
df = pd.read_csv("C:\\Users\\DIVISAVEE\\OneDrive\\Desktop\\Savee\\HeartAI\\data\\processed\\heart_cleaned.csv")

In [3]:
from sklearn.model_selection import train_test_split

X = df.drop("target", axis=1)
y = df["target"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [4]:
lr = LogisticRegression(
    max_iter=1000
)

lr.fit(X_train, y_train)

y_pred = lr.predict(X_test)

In [5]:
print("Accuracy :", accuracy_score(y_test, y_pred))
print("Recall    :", recall_score(y_test, y_pred))
print("Precision :", precision_score(y_test, y_pred))
print("F1 Score  :", f1_score(y_test, y_pred))

Accuracy : 0.8032786885245902
Recall    : 0.8484848484848485
Precision : 0.8
F1 Score  : 0.8235294117647058


## Logistic Regression Results

- Accuracy: 80.33%
- Recall: 84.85%
- Precision: 80.00%
- F1 Score: 82.35%

### Observation

Logistic Regression provides a strong baseline model with good recall. However, approximately 15% of heart disease cases remain undetected, making it unsuitable as the final model for a clinical decision support system.

In [6]:
rf = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

rf.fit(X_train, y_train)

rf_pred = rf.predict(X_test)

print("Accuracy :", accuracy_score(y_test, rf_pred))
print("Recall    :", recall_score(y_test, rf_pred))
print("Precision :", precision_score(y_test, rf_pred))
print("F1 Score  :", f1_score(y_test, rf_pred))

Accuracy : 0.7540983606557377
Recall    : 0.7878787878787878
Precision : 0.7647058823529411
F1 Score  : 0.7761194029850746


## Random Forest Results

- Accuracy: 75.41%
- Recall: 78.79%
- Precision: 76.47%
- F1 Score: 77.61%

### Observation

Random Forest underperformed compared to Logistic Regression. This suggests that the dataset may not benefit significantly from ensemble bagging methods due to its small size and relatively simple feature relationships.

In [9]:
xgb = XGBClassifier(
    n_estimators=200,
    max_depth=4,
    learning_rate=0.05,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42
)

xgb.fit(X_train, y_train)

xgb_pred = xgb.predict(X_test)

print("Accuracy :", accuracy_score(y_test, xgb_pred))
print("Recall    :", recall_score(y_test, xgb_pred))
print("Precision :", precision_score(y_test, xgb_pred))
print("F1 Score  :", f1_score(y_test, xgb_pred))

Accuracy : 0.7377049180327869
Recall    : 0.7272727272727273
Precision : 0.7741935483870968
F1 Score  : 0.75


"After evaluating multiple models, Logistic Regression demonstrated the best balance between simplicity, interpretability, and recall on this dataset."

In [11]:
for threshold in [0.5, 0.45, 0.4, 0.35]:
    pred = (lr.predict_proba(X_test)[:,1] >= threshold).astype(int)

    print(
        threshold,
        recall_score(y_test, pred)
    )

0.5 0.8484848484848485
0.45 0.8484848484848485
0.4 0.8787878787878788
0.35 0.9090909090909091


"To minimize false negatives, threshold tuning was applied to the Logistic Regression model. Reducing the classification threshold from 0.50 to 0.35 increased Recall from 84.85% to 90.91%, making the model more suitable for medical screening applications."

In [12]:
import joblib

joblib.dump(
    lr,
    "C:/Users/DIVISAVEE/OneDrive/Desktop/Savee/HeartAI/models/heart_model.pkl"
)

['C:/Users/DIVISAVEE/OneDrive/Desktop/Savee/HeartAI/models/heart_model.pkl']